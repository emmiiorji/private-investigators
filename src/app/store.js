import { configureStore } from '@reduxjs/toolkit';
import investigatorsReducer from './redux/investigator/investigatorSlice';

const store = configureStore({
  reducer: {
    investigators: investigatorsReducer,
  },
});

export default store;
