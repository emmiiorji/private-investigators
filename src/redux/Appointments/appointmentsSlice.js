/* eslint-disable no-param-reassign */
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export const fetchAppointments = createAsyncThunk('appointments/fetch', async () => {
  const response = await axios(`${process.env.REACT_APP_SERVER_BASE_URL}/v1/appointments/`, { withCredentials: true });
  return response.data;
});

export const deleteAppointment = createAsyncThunk('appointments/delete', async (appointmentId) => {
  await axios.delete(`${process.env.REACT_APP_SERVER_BASE_URL}/v1/appointments/${appointmentId}`, { withCredentials: true });
  return appointmentId;
});

export const createAppointment = createAsyncThunk('appointments/create', async (appointment) => {
  const response = await axios.post(`${process.env.REACT_APP_SERVER_BASE_URL}/v1/appointments/`, {
    appointment: {
      date: appointment.date,
      city: appointment.city,
      investigator_id: appointment.investigator_id,
    },
  }, { withCredentials: true });
  return response.data;
});

const initialState = {
  status: 'idle',
  value: [],
};

const appointmentsSlice = createSlice({
  name: 'appointments',
  initialState,
  reducers: {
    loadAppointments: (state, action) => {
      const newState = { ...state, value: [...state.value, action.payload] };
      return newState;
    },
  },

  extraReducers: (builder) => {
    builder
      .addCase(fetchAppointments.pending, (state) => {
        const newState = { ...state, loading: true };
        return newState;
      })
      .addCase(fetchAppointments.rejected, (state) => {
        const newState = { ...state, error: 'Error 404. Failed to fetch', loading: false };
        return newState;
      })
      .addCase(fetchAppointments.fulfilled, (state, action) => {
        const newState = { ...state, value: action.payload, status: 'fetched' };
        return newState;
      })
      .addCase(deleteAppointment.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(deleteAppointment.fulfilled, (state, action) => {
        state.status = 'deleted';
        state.value = state.value.filter(
          (appointment) => appointment.id !== parseInt(action.payload, 10),
        );
      })
      .addCase(deleteAppointment.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message;
      })
      .addCase(createAppointment.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.value = [...state.value, action.payload];
      });
  },
});

export default appointmentsSlice.reducer;
