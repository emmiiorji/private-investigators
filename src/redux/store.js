import { configureStore } from '@reduxjs/toolkit';
// import userReducer from './CurrentUser/CurrentUserSlice';
import investigatorsReducer from './investigator/investigatorSlice';
import appointmentsReducer from './Appointments/appointmentsSlice';
import userReducer from './CurrentUser/CurrentUserSlice';

const store = configureStore({
  reducer: {
    investigators: investigatorsReducer,
    appointments: appointmentsReducer,
    userReducer,
  },
});

export default store;
