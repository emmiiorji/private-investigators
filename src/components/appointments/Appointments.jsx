import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { nanoid } from '@reduxjs/toolkit';
import { deleteAppointment, fetchAppointments } from '../../redux/Appointments/appointmentsSlice';
import { fetchInvestigators } from '../../redux/investigators/investigatorSlice';

const Appointments = () => {
  const appointments = useSelector((state) => state.appointments);
  const investigators = useSelector((state) => state.investigators);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchAppointments());
    dispatch(fetchInvestigators());
  }, [dispatch]);

  const handleDelete = (event) => {
    dispatch(deleteAppointment(event.target.id));
  };

  return (
    <div>
      <h1>Appointments</h1>
      <table>
        <thead>
          <tr>
            <th>Investigator</th>
            <th>City</th>
            <th>Date</th>
            <th>Action</th>
          </tr>
        </thead>
        {
          investigators.status === 'fetched'
        && (
        <tbody>
          {appointments.value.map((appointment) => (
            <tr key={nanoid()}>
              <td>
                {investigators.value.find(({ id }) => id === appointment.investigator_id).name}
              </td>
              <td>{appointment.city}</td>
              <td>{appointment.date}</td>
              <td><button type="button" onClick={handleDelete} id={appointment.id}>Cancel</button></td>
            </tr>
          ))}
        </tbody>
        )

        }
      </table>
    </div>
  );
};

export default Appointments;
