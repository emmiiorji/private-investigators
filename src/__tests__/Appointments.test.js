import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/store';
import Appointments from '../components/appointments/Appointments';

describe('Appoinments tests', () => {
  let appointments;
  beforeAll(async () => {
    appointments = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Appointments />
          </Router>
        </Provider>,
      )
      .toJSON();
  });

  it('Component should render correctly', () => {
    expect(appointments).toBeTruthy();
  });

  it('It should match its snapshot', () => {
    expect(appointments).toMatchSnapshot();
  });

  test('It should render the component with the correct heading', () => {
    const heading = waitFor(() => screen.getByRole('heading', { name: 'Appointments' }));
    waitFor(() => expect(heading).toBeInTheDocument());
  });
});
