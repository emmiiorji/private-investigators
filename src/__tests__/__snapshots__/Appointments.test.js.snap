// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appoinments tests It should match its snapshot 1`] = `
<div>
  <h1>
    Appointments
  </h1>
  <table>
    <thead>
      <tr>
        <th>
          Investigator
        </th>
        <th>
          City
        </th>
        <th>
          Date
        </th>
        <th>
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Harry Gator
        </td>
        <td>
          Manhattan, NY
        </td>
        <td>
          2022-12-12T00:00:00.000Z
        </td>
        <td>
          <button
            id={1}
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Simon Gator
        </td>
        <td>
          Austin, TX
        </td>
        <td>
          2022-05-07T00:00:00.000Z
        </td>
        <td>
          <button
            id={2}
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
